Support Online Mode Login

Kas-tle opened the pull request:
This PR uses https://github.com/RaphiMC/MinecraftAuth to support online mode login when enabled in the config (disabled by default).

Kas-tle committed:
Initial attempt at login (not working)

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Online mode works

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Remove unused imports

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Don't bother with mojang pub key in offline mode

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Delete vscode settings

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Comments

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle commented:
Closes #13

IndianBartonka commented:
This need to login to Microsoft account via site?

IndianBartonka commented:
Nevermind

RaphiMC commented on src/main/java/org/cloudburstmc/proxypass/network/bedrock/session/UpstreamPacketHandler.java:
```diff
@@ -243,8 +243,8 @@ private void initOnlineLoginChain(StepMCChain.MCChain mcChain) throws Exception
             initMojangPublicKey();
         }
 
```
You should use a permalink here else the line numbers will point into random code when I update that class in the future

RaphiMC commented on src/main/java/org/cloudburstmc/proxypass/network/bedrock/session/UpstreamPacketHandler.java:
```diff
@@ -147,4 +182,88 @@ private void initializeProxySession() {
+            initMojangPublicKey();
+        }
+
+        // adapted from https://github.com/RaphiMC/ViaBedrock/blob/main/src/main/java/net/raphimc/viabedrock/protocol/packets/LoginPackets.java#L276-L291
```
You should use a permalink here else the line numbers will point into random code when I update that class in the future

Kas-tle replied:
https://github.com/CloudburstMC/ProxyPass/pull/47/commits/7018fa50bc1c273bbafba2089abfc5f3c542b709

RaphiMC commented:
My thoughts (Copied from discord conversation with @Kas-tle):

--- Skin Data issue ---
I noticed you didn't change the skin data contents when in online mode. You should at least change the username to the new one. But there are couple more things to change. For that reference ViaBedrock again
https://github.com/RaphiMC/ViaBedrock/blob/main/src/main/java/net/raphimc/viabedrock/protocol/providers/SkinProvider.java

And here is the code which creates the AuthChainData: https://github.com/ViaVersion/ViaProxy/blob/main/src/main/java/net/raphimc/viaproxy/proxy/external_interface/ExternalInterface.java#L92

--- Persistent account feature ---
Another nice thing you could implement is to save the tokens on disk so its not so annoying to have to relogin into the account everytime (Which could take some time depending on your password/MFA setup). MinecraftAuth already has builtin serialize/deserialize to json

--- JWT library issue ---
Avoid mixing JWT libraries. MinecraftAuth uses a different one then ProxyPass internally, but you could use the one from ProxyPass for all of your code instead of using the one from MinecraftAuth, so ProxyPass doesn't ever reference another JWT library in its code. I know it might be easier to use the copied snippets from ViaBedrock, but having better code consistency might be more future proof

Kas-tle committed:
Use jose4j for jwt

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
close http client

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Implement credential saving

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Correctly set skin data & refactor

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Use permalink

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle commented:
### Addressing comments

#### Skin Data Issue
Online mode skin data is now modified to use correct `PlayFabId`, `DeviceId`, `ThirdPartyName`, and `ServerAddress` per https://github.com/CloudburstMC/ProxyPass/pull/47/commits/dfa6c518e3f3d8a8dc55d977f99cf036363d6e76#diff-9431c08ad1b545817aa18196e72b3d80b0e03d4b316c610e0412de53459cd76aR107-R113.

#### Persistent Account Feature
Auth data is now saved to `auth.json` if `save-auth-details` is enabled in the config (defaults to `false`). On future logins, assuming the file is present, it is refreshed per https://github.com/CloudburstMC/ProxyPass/pull/47/files#diff-b7912aaad8a748c4b6abc5c22f42c328cf9bb3702f14b3f207ff8e0d15c68fa2R357-R362.

#### JWT Library Issue
Jose4j is now used for all JWTs per https://github.com/CloudburstMC/ProxyPass/pull/47/files#diff-9431c08ad1b545817aa18196e72b3d80b0e03d4b316c610e0412de53459cd76aR59-R85.

Kas-tle committed:
MOJANG_PUBLIC_KEY as constant

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle committed:
Update readme

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

RaphiMC reviewed (approved):
LGTM

Kas-tle committed:
Update default config

Kas-tle committed:
Merge branch 'CloudburstMC:master' into online-mode-java

Kas-tle committed:
Update MinecraftAuth due to public key change

Kas-tle committed:
Update mojang public key

Kas-tle committed:
Merge branch 'CloudburstMC:master' into online-mode-java

Kas-tle committed:
minecraftauth on mvn central (ty @RpahMC)

Kas-tle committed:
1.20.40

Signed-off-by: GitHub <noreply@github.com>

smartcmd commented:
When we merge this PR?

Kas-tle committed:
Upstream

Signed-off-by: Joshua Castle <26531652+Kas-tle@users.noreply.github.com>

Kas-tle commented:
Closing so I can more easily distribute my own builds via GH-actions since there doesn't seem to be much interest in merging this... and that will also give me the ability to do less savory things to make connecting to some servers more tolerable due to bugs in Protocol.
If anyone would like to try again in the future to get these changes merged feel free to use my code. I have archived a tag of the repo before I make any other changes here: https://github.com/Kas-tle/ProxyPass/tree/online-mode-pre-hard-fork

alvin0319 commented:
To be honestly, I think it would be better to ship online mode login on Protocol library instead of ProxyPass so that it could be utilized in many places like custom proxies.

SupremeMortal commented:
Online mode was something I wanted to ship with Protocol v3 (still possible since it hasn't released yet), so that would be my preference for instead of implementing into ProxyPass itself.
